pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/internal/RuleChainAnalyzer.java

```diff
@@ -96,6 +105,14 @@ public Expression visit(AxisExpression e) {
         return super.visit(e);
     }
 
+    @Override
+    public Expression visit(LazyExpression e) {
+        insideLazyExpression = true;
+        Expression result = super.visit(e);
+        insideLazyExpression = false;
```

Comment:
I'm only mentioning this, because this kind of pattern is common in the parser (maybe it's not relevant here). In the parser when we have an attribute like "inInterface", which must be true inside of some context, then it's usually necessary to save the previous state and restore it. Like
```java
boolean prevCtx = insideLazyExpression;
insideLazyExpression = true;
Expression result = super.visit(e);
insideLazyExpression = prevCtx;
```

Otherwise you risk overwriting the "true" value if the bit was already set when entering the method.

Now I don't know if nested lazy expressions are a thing in the saxon expression tree, but this may be a bug that comes up later to bite us

Reply:
Good point, I'll fix this